### PR TITLE
btc-spv: add difficulty calculation & headerchain validation

### PR DESCRIPTION
**Pull request by pathoam**
#### Problem
btc-spv api does not include difficulty calculation & headerchain validation

#### Summary of Changes
added it, and tests
Fixes #

### CODE REVIEW

File: programs/btc_spv_api/src/spv_processor.rs [hunk 1]
```diff
@@ -187,4 +214,27 @@ mod test {
...
+        for h in 0..header_chain_raw.len() {
+            let head = header_chain_raw[h];
+            println!("{}", head.0);
+            let header = BlockHeader::hexnew(head.0, head.1).unwrap();
+            header_chain.push(header);
+        }
+        let difficulty: u64 = 10_000_000;
+        let confirmations: u16 = 5;
+
+        SpvProcessor::validate_header_chain(header_chain, difficulty, confirmations);
```
Comment (garious) [inline programs/btc_spv_api/src/spv_processor.rs]: This call is failing.  Need to `unwrap()` this result.

File: programs/btc_spv_api/src/spv_state.rs [hunk 2]
```diff
@@ -87,10 +88,31 @@ impl BlockHeader {
...
+                    0,
+                )
+            } else {
+                (bits & 0x00FF_FFFF, 8 * ((bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative
+        if mant > 0x007F_FFFF {
+            Default::default()
```
Comment (garious) [inline programs/btc_spv_api/src/spv_state.rs]: Use `U256::default()` to fix the CI nit.

**Comment by pathoam** [comment]
I made the mistake of making this pr shortly before I left for india for 2 weeks. I don't have the resources and time at hand to effectively work on it right now, so consider it a wip until I get back. 
@garious

**Comment by garious** [comment]
Feel free to reopen when this is ready for a review.

**Pull request closed without merging**